### ISSUE TITLE
Please include hsdis in Adoptium builds

### ISSUE THREAD

**Issue by theRealAph** [adoptium/adoptium#213]
hsdis (a disassembler) is extremely useful when analysing HotSpot performance. At present Temurin builds don't support hsdis.

Capstone is a "univeral disassembler" project. It is easy to use Capstone to build OpenJDK's hsdis.
To build OpenJDK using Capstone you need to have Capstone installed. Typical ways of installation can be `sudo apt install libcapstone-dev` (on Debian and derivatives), or brew install capstone (on macOS with Homebrew). For Windows, you need to download the "Core Engine", and unzip it. See the [Capstone Download page](https://www.capstone-engine.org/download.html#windows---core-engine-) for up-to-date download links.

The easiest way to build with support for hsdis on RHEL/CentOS is something like this:

```
sudo dnf -y install capstone-devel
sh ./configure ... --enable-hsdis-bundling --with-hsdis=capstone
```

After this the newly-built shared library `hsdis-XX.so` (not the built JDK, just hsdis) has a _runtime_ dependency on` libcapstone.so.4.x`. So, not only must a user install that library, but they must install Version 4. This is rather unfortunate. 

An alternative would be to link hsdis _statically_ against libcapstone. This means that we have no runtime dependency on capstone. This would require me to make a small change to the OpenJDK configuration upstream.

As far as I'm aware there are no licence problems using Capstone in this way.

**Comment by tellison** [comment]
Thanks for opening the discussion.

Just to note that `hsdis` [source code](https://github.com/openjdk/jdk/tree/master/src/utils/hsdis) is provided under "The Universal Permissive License (UPL), Version 1.0"

**Comment by tellison** [comment]
@theRealAph Reading a bit around this tool for the first time, a popular backend option is to use `--with-hsdis=binutils`. Was there a reason to suggest using [Capstone](https://www.capstone-engine.org/) rather than [Gnu Binutils](https://www.gnu.org/software/binutils/)?

For the record, the resulting `hsdis` plug-in binary is approx. 2.2Mb.

**Comment by gnu-andrew** [comment]
The Binutils backend is the one that has been in OpenJDK since launch - and before there were different backends - and suffers from a license incompatibility between it (GPLv3+) and HotSpot (GPLv2). That's one of the main reasons the other backends were introduced.

There is also now an LLVM backend which is what I'm planning to look into packaging in RHEL & Fedora, as they already have LLVM packages available.

**Comment by theRealAph** [comment]
> @theRealAph Reading a bit around this tool for the first time, a popular backend option is to use `--with-hsdis=binutils`. Was there a reason to suggest using [Capstone](https://www.capstone-engine.org/) rather than [Gnu Binutils](https://www.gnu.org/software/binutils/)?
> 
> For the record, the resulting `hsdis` plug-in binary is approx. 2.2Mb.

From a technical point of view I'd use Binutils, but there's no licence-conforming way to to do it, as far as I can see from GPL v3. But it's something of a legal grey area. I don't want to go there.

Hsdis built with Capstone is about the same size as Binutils, at 2.4M.

**Comment by sxa** [comment]
Noting that CentOS6 (which we use for building on x86 Linux for JDK17 and blow)  does not have capstone in its repositories so we may need to build from source on there.

**Comment by sxa** [comment]
Noting that [openjdk only supports hsdis+capstone on x64 and aarch64](https://github.com/adoptium/jdk21/blob/c86f4dea9529640cd3234c5cad2f36f3201b1385/make/Hsdis.gmk#L45)

**Comment by sxa** [comment]
My current initial approach has been to install a suitable capstone as part of the machine setup process (within our ansible scripts) and then invoke the appropriate options during the openjdk build to pick it up. This should allow it to work on CentOS6 without the problems from the earlier comment, and also keep the library smaller (some default installations of capstone appear to contain support for every architecture in the library)

A second option would be to build (and probably therefore also download during each build) the library as part of the Temurin build process, so we'd rebuild it each time. Since this implementation is re-shipping the built library, this might be a preferable option. Thoughts?

**Comment by smlambert** [comment]
3rd option: 
- Download as part of a dependencies job on our Jenkins server and pull it from lastSuccessfulArtifacts of that job, removes 3rd party flakiness, reduces chances of network issues and gets you a mechanism for pulling other versions (if later that is a requirement).

**Comment by sxa** [comment]
After a couple of discussions today I am now leaning more towards an option 2.5 (Somewhere between 2 and 3) of pulling from pre-built binaries on stored in the `userContent` path on the jenkins server (Possibly similar to how we have a cache of artefacts like gcc at present) and downloading them onto the machine during the build to reduce dependencies on external servers.
Option 2 isn't ideal since it wouldn't automatically guarantee reproducibility of the build.

**Comment by theRealAph** [comment]
The main advantage of Capstone is that it is small, and it has a BSD licence, thus GPL v2 compatible. 
I'd like hdsis's footprint to be so small that people don't need it to be packaged separately. It is a fantastically-useful diagnostic tool.

Be aware that Capstone can be built in multiple ways. By default, it builds for *all* architectures. This is not what we want.
[](https://github.com/capstone-engine/capstone/blob/next/COMPILE.TXT)

**Comment by sxa** [comment]
> Be aware that Capstone can be built in multiple ways. By default, it builds for all architectures. This is not what we want.

Yep the [draft PR I had](https://github.com/adoptium/infrastructure/pull/3119) was to only build for the specific architecture. The approach I'm trialing after my last comment is to store prebuilt versions for each architecture in a form we can download and use within the build process to ensure that we are using an identical binary for each build for reproducibility comparison purposes.

Broadly speaking I'm doing this to build the version, which we will cache and use in a build PR I'm about to look at putting together :-)
```
export CC=/usr/local/gcc11/bin/gcc-11.2
export CXX=/usr/local/gcc11/bin/g++-11.2
wget https://github.com/capstone-engine/capstone/archive/4.0.2.tar.gz
tar xpfz 4.0.2.tar.gz
cd capstone-4.0.2/
export CAPSTONE_ARCHS=x86
export PREFIX=/usr/local/capstone-4.0.2
./make.sh
./make.sh install
tar czf capstone-4.0.2-x86_64.tar.gz -C /usr/local capstone-4.0.2
```

**Comment by sxa** [comment]
Test builds of jdk-21+35 with this support added have been built for [x64](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21/job/jdk21-linux-x64-temurin/120/artifact/workspace/target/) and [aarch64](https://ci.adoptium.net/job/build-scripts/job/jobs/job/jdk21/job/jdk21-linux-aarch64-temurin/97/artifact/workspace/target) (although they will not be retained there indefinitely). As a one-off for people to experiment with, I've made available a capstone download for a suitable 4.0.2 alongside the builds in those links.

**Comment by sxa** [comment]
The PMC decided today to not include it in the 20.0.1 October deliverable and to subsequently evaluate whether this was useful in a non-development build of the JDK. As an interim we can let users use the builds in the previous comment for evaluation purposes and feedback, or we can look at providing a separate hsdis library that people could add in if desired. This may be more useful once we get a statically linked version of capstone included into hsdis.

**Comment by theRealAph** [comment]
> The PMC decided today to not include it in the 20.0.1 October deliverable and to subsequently evaluate whether this was useful in a non-development build of the JDK.

This is the JDK, not the JRE, right? So it's for Java developers. Some popular tools, especially JMH, need hsdis in order to provide full functionality. Sure, hsdis can be built externally, but you need JDK source to do it, and our Java developers shouldn't have to do that, and many of them don't even know how to do that.

Being able to see the generated code is very useful as an educational tool for helping Java developers understand performance issues. C/C++ compilers provide visible assembly-language, and IMVHO all Java developers have a right to this information too.

**Comment by jerboaa** [comment]
My understanding was that the current approach of building `hsdis.so` was to **dynamically** link to `libcapstone.so`. That means even if the lib is included, users would still need to have `libcapstone.so` installed. It would probably be best if that step wouldn't be needed. If we want to include it, there should be an automated test verifying that it works as expected. Could you please provide us some use-cases that you think Java developers would use this for on release builds? Thanks!

**Comment by theRealAph** [comment]
> My understanding was that the current approach of building `hsdis.so` was to **dynamically** link to `libcapstone.so`.

I guess it would do, but that's worse from a user's perspective.

> That means even if the lib is included, users would still need to have `libcapstone.so` installed. It would probably be best if that step wouldn't be needed. If we want to include it, there should be an automated test verifying that it works as expected. Could you please provide us some specific use-cases (with examples perhaps) that you think Java developers would use this for on release builds? Thanks!

Use case is any JMH test with `-prof perfasm`. But verifying the output would be hard.

Maybe it would be sufficient to run the JVM with `-XX:+PrintAssembly` and make sure that the error messages that would be printed if the library was missing were not present.

**Comment by sxa** [comment]
> > The PMC decided today to not include it in the 20.0.1 October deliverable and to subsequently evaluate whether this was useful in a non-development build of the JDK.

> This is the JDK, not the JRE, right? So it's for Java developers. Some popular tools, especially JMH, need hsdis in order to provide full functionality

@tellison From what I recall (and it was some time ago now so forgive me if I've misremembered) you had some specific concerns with integrating the hsdis+capstone support into the Temurin build+distribution process and whether it was useful to people who did not have a full development environment. Based on the comments above there are situations in which this would be useful to a JDK user who was not an openjdk developer. Do you still have reservations about adding some or all of this functionality to support hsdis+capstone on a subset of our platforms?

**Comment by sxa** [comment]
_(Copying this in from something I posted in another email thread)_

It looks like we might have some issues relating to capstone versioning. A build of the hsdis module done against capstone 4 isn't compatible with capstone 5.0.1 as supplied with Fedora 40+ (I tried symlinking the .so..5 to .so.4 to help it link and it ran but the output wasn't usable) which means it's going to be tricky if we don't ship with capstone to provide something that works universally. My pre-built 4.0.2 worked ok on Fedora 40 though so one option would be to recommend that users running distributions which ship 5 by default "build capstone 4 like this ...". Or another option would be just support 5 going forward (I haven't built one with 5 yet to verify that).

A quick look at what's in each distributions repository (Note I haven't specifically verified that they work with my hsdis library at this point, other than the EL8 one):
systems
- EPEL for EL8 and EL9 seem to have 4.0.2
- All Ubuntu releases from 22.04 through 24.10 have 4.0.2 (20.04LTS is still in support and has a weird transitional package by the look of it)
- Fedora bumped the version between F39 and F40 (CentOS Stream 10 also has 5.0.1 so distributions based on that such as RHEL10 should have capstone 5)
- The current Debian releases (including "unstable" and "testing") seem to have 4.0.2

So an hsdis build against capstone 4.0.2 is likely to work with most current distributions, although since Fedora have bumped up to 5 it seems quite likely that others will in the next year too. It doesn't feel like a good idea to go down the route of only supporting 5 when no current LTS distributions have that in their repositories.

_FYI @shipilev since we were chatting about this_

**Comment by jerboaa** [comment]
Yes. @shipilev proposed to have hsdis bundling with upstream OpenJDK builds on the Feb 3, 2025 OpenJDK committers workshop. If that gains any traction this effort here might become obsolete. Certainly something worth keeping both communication channels open.

**Comment by shipilev** [comment]
We don't really have a concrete plan to bundle hsdis, so far. 

It looks like Capstone comes with compatibility problems that @sxa identified above. So it looks to me we would need to see if either: a) Capstone static build is possible to avoid runtime dependencies; or b) Doing LLVM backend on all platforms is possible, without complicating build configurations all too much.

**Comment by solonovamax** [comment]
> Could you please provide us some specific use-cases (with examples perhaps) that you think Java developers would use this for on release builds? Thanks!

I have multiple jdks installed on my device which I use for development, and the jdks I primarily use during development are the adoptium & jetbrains jdks (via sdkman)
There are times when I find it useful to be able to inspect how the jvm decides to compile things for analyzing performance, and ideally I would not need to switch to a different jdk to do this.

When I currently need to load hsdis, I just use arch's build of openjdk with hsdis I compiled using [java21-openjdk-hsdis](https://aur.archlinux.org/packages/java21-openjdk-hsdis), however if hsdis were included in adoptium this would simplify things a lot for me.

> My understanding was that the current approach of building `hsdis.so` was to dynamically link to `libcapstone.so`. That means even if the lib is included, users would still need to have `libcapstone.so` installed.

so long as this is documented somewhere on the website, I would have no issue with this. In my usecase, I would search something along the lines of "adoptium hsdis", see the page on the website about hsdis requiring libcapstone, and then install it on my system.

> I'd like hdsis's footprint to be so small that people don't need it to be packaged separately.

honestly, I'd even be quite fine with hsdis being packaged separately, so long as tools like sdkman supported it and package repos for different linux distros which package adoptium all packaged it.
perhaps packaging it separately might also allow for using the binutils backend? however, I am not a lawyer, so I have no clue if this would actually get around the license restrictions or not
